### DISCUSSION TITLE
.mat transfer function

### DISCUSSION THREAD

**Discussion by daniell-00** [bachlab/PsPM#654] [Q&A] [answered]
Hello, could someone share an example to create a .mat file containing transfer function for SCR? Thanks in advance.

Daniel

**Comment by dominikbach** [comment]


**Reply by daniell-00** [reply]
Hi Dominik,

I'll try to give you more context with some examples. I import skin conductance signal using graphic interface or the terminal and when I check it with the display tool I got this:

![Captura de pantalla 2024-03-18 160958](https://github.com/bachlab/PsPM/assets/153566537/b307bf8a-e9cf-4b33-b3a7-809b2ae57efd)


My data originally seems quite different, they look like this:

![imagen](https://github.com/bachlab/PsPM/assets/153566537/90ee0e0f-5eda-4396-8450-27ae69315c55)


I checked the manual trying to find some explanation and I thought that it may due to a missing of the function transfer when I selected options for importing:

![Captura de pantalla 2024-03-18 162621](https://github.com/bachlab/PsPM/assets/153566537/822aac30-547a-4ace-aa1c-983f1cfb423d)


So I was trying to test this supposition but I'm not sure how to specify the transfer function manually or using a .mat file.

Could you help me to clarify this issue? Thanks in advance.

Daniel

**Comment by dominikbach** [comment]
Hi Daniel

the transfer function defines a monotonic mapping from raw to imported data. From the screenshots, the problem seems to be something different. Also, there might be something wrong with the time representation of the data (original is 750 s long, imported data > 3500 s). We don't have a lot of cnt sample data, so the import function might not cover your specific case. If you send us a sample file (even just a short snippet) we are happy to investigate.

Dominik

**Comment by daniell-00** [comment, marked as answer]
Hi Dominik,

I imported data (same as .cnt) from .edf file and I it solved the problem. It seems there is an issue importing from .cnt format, additionally, the time is twice the original length (1966 s):

![Captura de pantalla 2024-03-19 152833](https://github.com/bachlab/PsPM/assets/153566537/074d3b58-e7cf-4b1d-9b2f-25d9a89a4bc5)

I assume it is better to use .edf format. On the other hand, I would like to know if there are any suggestions regarding the epoch length. I'll try to analyze skin conductance level on periods about 11 min using SF model.

P.D. I couldn't attached the .cnt file here.

Best regards,
Daniel

**Comment by dominikbach** [comment]
That looks good. And yes, there should be no problems using 11 minute epochs (although they might take 10-20 mins to invert, depending on the power of your computer). 

If you would be willing to send a cnt example file we'd appreciate this - you can use the email address in the PsPM start message.

**Comment by daniell-00** [comment]
Sorry, I couldn't manage myself to find the email address to share some .cnt files with you, could you share it with me please.

On the other hand I couldn't create the transfer function file because sample rate is bigger than 90, I don't know if it has an effect over the results when I run the functions for skin conductance level and skin conductance response using spontaneous fluctuations functions, could you help me to clarify it please. Additionally, I would like to know if there's a problem since in the example I shared with you the amplitude of data range is about -6500 to +6500.

Thank you so much!!
Daniel

**Comment by dominikbach** [comment]
To send files, you can use the email address in the PsPM startup message (in the command window).

The sample rate is not needed for the transfer function. 

The sample rate has no influence on the results, if it is properly specified during the import.

The amplitude range indicates that the data are not properly imported into µS units. To assist, we need some information on the recording system and the transfer function you are currently using.

**Comment by daniell-00** [comment]
Hello,

Following your advice I' tried to make a transfer function file as follows:

sr = 20; % Preferred sample rate
fn = 'transfer_function.mat'; % Transfer function file name
[y, t] = pspm_bf_scrf_f(sr);
save(fn, 'y', 't');
plot(t, y);


%% I got this plot:
![imagen](https://github.com/bachlab/PsPM/assets/153566537/9109bebf-6915-48eb-b422-a556598dff64)


%% I tried to run just importing function with transfer function file created previously: 

options = struct('overwrite', true);
epochs = 'eventos_2.mat';
import = {...
    struct('type', 'scr', 'channel', 5, 'transfer', 'transfer_function.mat'), ...
    struct('type', 'marker') ...
    };

for sujeto = 2
    pspm_import(sprintf('%i.edf',sujeto), 'edf', import, options);
end



%% I got the next error message:

» Importing 2.edf: Warning: /nNo conversion constant given 
Output argument "data" (and possibly others) not assigned a value in the execution with "pspm_get_scr" function.

Error in pspm_import (line 271)
      [sts(k), data{k}] = feval(settings.channeltypes(channeltype).import, import{blk}{k});

Error in importar (line 21)
    pspm_import(sprintf('%i.edf',sujeto), 'edf', import, options); 

%% I'm using a g.tec device for registering the GSA:

![IMG_20230214_114438](https://github.com/bachlab/PsPM/assets/153566537/851e8b26-6778-4c53-b0be-60073c62cf11)



%% I'm using the script bellow to process the data, I don't specify the sample rate which is 250 hz, I thought it was automatically detected:

import = {...
    struct('type', 'scr', 'channel', 5, 'transfer', 'none'), ...
    struct('type', 'marker') ...
    };
                                                     
for sujeto = 1:2
    epochs = (sprintf('epocas_%i.mat',sujeto));
    options = struct('overwrite', true);
    pspm_import(sprintf('%i.edf',sujeto), 'edf', import, options);       
    model = struct('datafile', sprintf('pspm_%i.mat', sujeto), 'modelfile', sprintf('msf_%i', sujeto), 'timing', epochs, 'timeunits', 'seconds', 'method', 'dcm');
    pspm_sf(model, options);
    options = struct('target', sprintf('st_%i', sujeto));
    pspm_exp(sprintf('msf_%i', sujeto), options);
end


I'll be so glad of reading your suggestions. Thank you so much for your support!!!

P.D. I've sent some .cnt files

Daniel

**Comment by dominikbach** [comment]
Hi Daniel. 

My first post was confusing - I had misunderstood your question - and I have deleted it, so that others don't get confused too.

A transfer function file for SCR import is created in the following way.

```
c = ...
Rs = 0;
offset = 0;
recsys = 'conductance';
save('transfer.mat', 'c', 'Rs', 'offset', 'recsys');
```
From the information you posted, your system records 6.7 mV per 1 µS conductance change. If you are recording your data in units of "mV", then you should set `c = 6.7;` in the script above. If you are recording in units of "V" then it should be `c = 6.7/1000;`

Hope this makes more sense. You can get detailed help by typing `help pspm_transfer_function`.
Dominik

**Comment by daniell-00** [comment]
Hi Dominik,

I did as you suggested. I would like to verify if the graphics I got using the transfer function have any sense. I'm concerned about this because I don't know if values below cero are a problem for the results I had using the PsPM sf functions. I add some graphs as an example:

![Captura de pantalla 2024-04-12 164926](https://github.com/bachlab/PsPM/assets/153566537/8f2acef1-ec0a-4aa4-8723-271e9fcb3268)

![Captura de pantalla 2024-04-12 164823](https://github.com/bachlab/PsPM/assets/153566537/1b8a5922-98ae-4c07-9411-85d0fb6fe9b1)

![Captura de pantalla 2024-04-12 163054](https://github.com/bachlab/PsPM/assets/153566537/ba36e905-73fe-4fe7-8897-d2ce0991ae87)

Thanks in advance.

Daniel

**Reply by dominikbach** [reply]
Hi Daniel.

1. Did you use `c = 6.7` or `c = 6.7/1000`? It seems your values are on a factor 1000 larger than they should be, so I suspect you used the former identity when it should be the latter. (From the second plot you posted in this thread, it also seems your system records in mV indeed, not in V). 
2. Skin conductance cannot be negative by definition; however you have negative values even in your original data, as your second plot in the thread shows [(here)](https://private-user-images.githubusercontent.com/153566537/313851728-90ee0e0f-5eda-4396-8450-27ae69315c55.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMxOTA4OTMsIm5iZiI6MTcxMzE5MDU5MywicGF0aCI6Ii8xNTM1NjY1MzcvMzEzODUxNzI4LTkwZWUwZTBmLTVlZGEtNDM5Ni04NDUwLTI3YWU2OTMxNWM1NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQxNVQxNDE2MzNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jMDBhOTE1NDQ1ZWJjNDA3OTRhMjZhZmRjYjgwZmU1NTRjZjBlNGNiMzhlODJlYTAwZjViYmQ1OThhMGVhOTdhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.g0U7ZwjUnTxo3oToaobTQKEh-Sv1lbDazOvwLKfxaxI) You might want to check your recording system.

**Reply by daniell-00** [reply]
Hi Dominik,

Yes, I used c = 6.7. I'll change it for c = 6.7/1000. On the other side, the amplifier is transforming the data to a range of negative and positive values and is a default configuration option which is not possible to change, so I would like to know if there is an option to normalize/rescale the data using PsPM? Thanks in advance.

Best wishes
Daniel

**Comment by dominikbach** [comment]
Hi Daniel

you only need the absolute SCR units if you are assessing SF frequency - because they require an absolute threshold. There are other ways of analysing SF data, for example extracting the amplitude of each response and using it directly (without thresholding). Also, no absolute SCR units are required for GLM or the non-linear model.

For all these latter models, it is indeed possible to normalise the data (and this is advisable for within-subjects designs), using the setting `model.norm`.  

Dominik
